### PR TITLE
Support RLOO

### PR DESCRIPTION
**Pull request by zhuzilin**
This PR brings in RLOO (REINFORCE Leave-One-Out), using an approach similar to the one in [TRL](https://github.com/huggingface/trl/blob/0238d96c6f43abd808804ddb0065883964073060/trl/trainer/rloo_trainer.py#L386).

One thing to note: I didn’t change `experience.info["reward"]` in place. This way, wandb still logs the actual reward values, which makes it easier to compare results across different algorithms.

Thank you for your time on reviewing this PR:)

### CODE REVIEW

**Comment by hijkzzz** [comment]
@wwxFromTju helps review this first.

**Comment by wwxFromTju** [comment]
I looked at the code and it was clearly written. Further, I examined the computational part of RLOO, LGTM!

However, I need time to further compare it with other open source implementations so that I can determine if there is a better implementation to suggest.

File: openrlhf/trainer/ppo_utils/experience_maker.py [hunk 1]
```diff
@@ -330,9 +330,23 @@ def make_experience(self, samples: Samples) -> Experience:
...
+        Process experiences, this can be used to filter out some experiences or do some processing on the rewards.
+
+        Output:
+        - experiences: List of Experience
+        - rewards: List of rewards
+        """
+        args = self.strategy.args
+        if args.advantage_estimator == "rloo":
+            rewards = torch.cat([experience.info["reward"] for experience in experiences])
+            rewards = rewards.reshape(args.n_samples_per_prompt, -1)
```
[the remark this answers is not in the document]
Comment (zhuzilin) [reply openrlhf/trainer/ppo_utils/experience_maker.py]: oh... yes, it should be `reshape(-1, args.n_samples_per_prompt)`. fixed.

[the remark this answers is not in the document]
Comment (wwxFromTju) [reply openrlhf/trainer/ppo_utils/experience_maker.py]: OH, I'm also confusing it with the other way of generating it. According to this generation, https://github.com/OpenRLHF/OpenRLHF/blob/0d5d2f9505fd03cf6976833343e4e96d9ac9513d/openrlhf/trainer/ppo_utils/experience _maker.py#L244, the reshape method we need to change

**Comment by hijkzzz** [comment]
55a9b371a87dfe7c1dac8449904a54c

**Comment by wwxFromTju** [comment]
It doesn't look like grpo is very effective, and I'm inclined to merge cautiously.

**Comment by hijkzzz** [comment]
A question: how rloo supports PRM?

**Comment by songxxzp** [comment]
Maybe there is a difference between the TRL RLOO implementation and this PR.

In TRL:
```python
class PPOTrainer(Trainer):
    def train(self):
                ...
                # 4. compute rewards
                kl = logprobs - ref_logprobs
                non_score_reward = (-args.kl_coef * kl).sum(1)
                rlhf_reward = scores + non_score_reward

                # vectorized RLOO advantages implementation
                rlhf_reward = rlhf_reward.reshape(args.rloo_k, -1)
                baseline = (rlhf_reward.sum(0) - rlhf_reward) / (args.rloo_k - 1)
                advantages = rlhf_reward - baseline
                advantages = advantages.flatten()
                torch.cuda.empty_cache()
```
In TRL, all KL rewards are summed for all tokens. The baseline is computed after KL.

RLOO

In this PR, the baseline is computed first, then added with KL, but they are not summed together:
```python
class NaiveExperienceMaker(ABC):
    ...
    @torch.no_grad()
    def make_experience_list(self, all_prompts: Union[str, List[str]], **generate_kwargs) -> List[Experience]:
        ...
        experiences, rewards = self.process_experiences(experiences)

        # calculate return and advantages
        for experience, reward in zip(experiences, rewards):
            num_actions = experience.info["num_actions"]
            reward = compute_reward(
                reward,
                self.kl_ctl.value,
                experience.kl,
                action_mask=experience.action_mask,
                num_actions=num_actions,
                reward_clip_range=args.reward_clip_range,
            )
        ...

    @torch.no_grad()
    def process_experiences(self, experiences: List[Experience]) -> Tuple[List[Experience], List[torch.Tensor]]:
        """
        Process experiences, this can be used to filter out some experiences or do some processing on the rewards.

        Output:
        - experiences: List of Experience
        - rewards: List of rewards
        """
        args = self.strategy.args
        if args.advantage_estimator == "rloo":
            rewards = torch.cat([experience.info["reward"] for experience in experiences])
            rewards = rewards.reshape(-1, args.n_samples_per_prompt)
            baseline = (rewards.sum(-1, keepdim=True) - rewards) / (args.n_samples_per_prompt - 1)
            rewards = rewards - baseline
            rewards = rewards.flatten().chunk(len(experiences))
            return experiences, rewards
        return experiences, [experience.info["reward"] for experience in experiences]
```
```python
def compute_reward(...):
    ....
    if action_mask is not None:
        kl_reward = -kl_coef * kl
        # The following code is equivalent to:
        #
        # last_reward = torch.zeros_like(kl)
        # for i in range(last_reward.size(0)):
        #     for t in reversed(range(last_reward.size(1))):
        #         if action_mask[i][t] > 0.5:
        #             last_reward[i][t] = r[i]
        #             break
        #
        eos_indices = action_mask.size(1) - 1 - action_mask.long().fliplr().argmax(dim=1, keepdim=True)
        last_reward = torch.zeros_like(kl).scatter_(dim=1, index=eos_indices, src=r.unsqueeze(1).to(kl.dtype))

        reward = last_reward + kl_reward
    else:
        # TODO: write a more efficient version
        reward = []
        for i, (kl_seg, action_len) in enumerate(zip(kl, num_actions)):
            kl_reward = -kl_coef * kl_seg
            kl_reward[action_len - 1] += r[i]
            reward.append(kl_reward)
```

**Comment by hijkzzz** [comment]
> Maybe there is a difference between the TRL RLOO implementation and this PR.
> 
> In TRL:
> 
> ```python
> class PPOTrainer(Trainer):
>     def train(self):
>                 ...
>                 # 4. compute rewards
>                 kl = logprobs - ref_logprobs
>                 non_score_reward = (-args.kl_coef * kl).sum(1)
>                 rlhf_reward = scores + non_score_reward
> 
>                 # vectorized RLOO advantages implementation
>                 rlhf_reward = rlhf_reward.reshape(args.rloo_k, -1)
>                 baseline = (rlhf_reward.sum(0) - rlhf_reward) / (args.rloo_k - 1)
>                 advantages = rlhf_reward - baseline
>                 advantages = advantages.flatten()
>                 torch.cuda.empty_cache()
> ```
> 
> In TRL, all KL rewards are summed for all tokens. The baseline is computed after KL.
> 
> RLOO
> 
> In this PR, the baseline is computed first, then added with KL, but they are not summed together:
> 
> ```python
> class NaiveExperienceMaker(ABC):
>     ...
>     @torch.no_grad()
>     def make_experience_list(self, all_prompts: Union[str, List[str]], **generate_kwargs) -> List[Experience]:
>         ...
>         experiences, rewards = self.process_experiences(experiences)
> 
>         # calculate return and advantages
>         for experience, reward in zip(experiences, rewards):
>             num_actions = experience.info["num_actions"]
>             reward = compute_reward(
>                 reward,
>                 self.kl_ctl.value,
>                 experience.kl,
>                 action_mask=experience.action_mask,
>                 num_actions=num_actions,
>                 reward_clip_range=args.reward_clip_range,
>             )
>         ...
> 
>     @torch.no_grad()
>     def process_experiences(self, experiences: List[Experience]) -> Tuple[List[Experience], List[torch.Tensor]]:
>         """
>         Process experiences, this can be used to filter out some experiences or do some processing on the rewards.
> 
>         Output:
>         - experiences: List of Experience
>         - rewards: List of rewards
>         """
>         args = self.strategy.args
>         if args.advantage_estimator == "rloo":
>             rewards = torch.cat([experience.info["reward"] for experience in experiences])
>             rewards = rewards.reshape(-1, args.n_samples_per_prompt)
>             baseline = (rewards.sum(-1, keepdim=True) - rewards) / (args.n_samples_per_prompt - 1)
>             rewards = rewards - baseline
>             rewards = rewards.flatten().chunk(len(experiences))
>             return experiences, rewards
>         return experiences, [experience.info["reward"] for experience in experiences]
> ```
> 
> ```python
> def compute_reward(...):
>     ....
>     if action_mask is not None:
>         kl_reward = -kl_coef * kl
>         # The following code is equivalent to:
>         #
>         # last_reward = torch.zeros_like(kl)
>         # for i in range(last_reward.size(0)):
>         #     for t in reversed(range(last_reward.size(1))):
>         #         if action_mask[i][t] > 0.5:
>         #             last_reward[i][t] = r[i]
>         #             break
>         #
>         eos_indices = action_mask.size(1) - 1 - action_mask.long().fliplr().argmax(dim=1, keepdim=True)
>         last_reward = torch.zeros_like(kl).scatter_(dim=1, index=eos_indices, src=r.unsqueeze(1).to(kl.dtype))
> 
>         reward = last_reward + kl_reward
>     else:
>         # TODO: write a more efficient version
>         reward = []
>         for i, (kl_seg, action_len) in enumerate(zip(kl, num_actions)):
>             kl_reward = -kl_coef * kl_seg
>             kl_reward[action_len - 1] += r[i]
>             reward.append(kl_reward)
> ```

Per-token KL can be used for early credit assignment, and TRL's implementation is not necessarily authoritative....

**Comment by hijkzzz** [comment]
I get your point, but essentially, this is just an RLOO implementation for more convenient coding. I believe that per-token KL is a more reasonable approach in RLHF.

**Comment by songxxzp** [comment]
I'm also interested in this question. RLOO only considers one-step conditions; multi-step conditions are not specified. However, I guess we can use $R(s_i) = sum(KL_{s_i}) + reward(S_i)$ as the $Process Reward(s_i)$ for the step $s_i$, refering to MATH-SHEPHERD. (Or maybe $sum(R(s_1), R(s_2), ..., R(s_i))$ could be used, who knows?)

When calculating the baseline, maybe we can use all the $Process Reward$ values from all steps in all responses except the current step, or perhaps use the $Process Reward$ from other responses? I believe one key point is how the advantage or baseline is calculated.

By the way, regarding per-token KL, I also believe that dense and detailed feedback could be helpful. However, as the reward is "sparse," RLOO seems to have "made a compromise" to maintain consistency in KL. I think that when token-level PRM is implemented, per-token KL could demonstrate its best advantages with token-level rewards.

**Comment by hijkzzz** [comment]
The summation  `sum ( R ( s1 ) , R ( s2 ) , . . . , R ( si ) )`  would render PRM meaningless. The basic version of RLOO is inherently incompatible with PRM. Currently, I am inclined to use GRPO or RLOO as a component of reward shaping rather than for advantage computation; otherwise, they cannot be well integrated with PRM.

**Comment by hijkzzz** [comment]
> @hijkzzz I am convinced(about KL). However, I suggest clarifying the differences from the implementation in the RLOO paper in future versions. From personal experience, I implemented multimodal content in the llama-factory's PPO, but the performance was consistently poor. I eventually discovered that its PPO algorithm is completely different from OPENAI's RLHF PPO. This is also why I started wandering through various alignment repositories...
> 
> A possibly unrelated question, does OpenRLHF have plans to add multimodal support in the future?


In fact, I think the official implementation of RLOO and GRPO is not particularly elegant from an RL perspective. OpenRLHF places great emphasis on implementation rigor, which is why we've been hesitant about how to incorporate RLOO. In our testing, we haven’t found any performance advantages of RLOO/GRPO over REINFORCE.....

Multi-modal PPO is complex story. Just added it to the TODO list.

**Comment by hijkzzz** [comment]
> @hijkzzz
> 
> > 55a9b371a87dfe7c1dac8449904a54c
> 
> I just made a simple implementation of TRL's RLOO in OpenRLHF. Could you share your experiment settings (or configuration)? I'd like to examine the gap between PPO, REINFORCE, and RLOO within OpenRLHF.

1. KL coef = 0.001 (a small value, so that how to implement KL reward is not important)
2. Experiments 1: MetaMathQA datasets + rule-based reward
3. Experiments 2: our default datasets and reward model
4. max_samples = 25000, n_samples = 4

**Comment by songxxzp** [comment]
@hijkzzz 
Thanks for sharing. I'm currently running PPO using the Experiment 2 settings(on main branch).
https://wandb.ai/songxxzp/openrlhf_train_ppo?nw=nwusersongxxzp
Screenshot_20241203_220103

However, the reward is constantly oscillating, which makes me worry that I might have set the parameters incorrectly, such as the learning rate and batch size. Here is my configuration (I've increased n_samples_per_prompt to 8):
Here is my configuration (I've increased n_samples_per_prompt to 8). Could there be any significant misalignment?"
```
openrlhf.cli.train_ppo \
   --pretrain OpenRLHF/Llama-3-8b-sft-mixture \
   --reward_pretrain OpenRLHF/Llama-3-8b-rm-mixture \
   --save_path ./checkpoint/llama-3-8b-ppo-64-8x8 \
   --save_steps -1 \
   --logging_steps 1 \
   --eval_steps -1 \
   --micro_train_batch_size 2 \
   --train_batch_size 64 \
   --micro_rollout_batch_size 4 \
   --rollout_batch_size 8 \
   --max_epochs 1 \
   --prompt_max_len 1024 \
   --generate_max_len 1024 \
   --zero_stage 2 \
   --bf16 \
   --actor_learning_rate 5e-7 \
   --critic_learning_rate 9e-6 \
   --init_kl_coef 0.001 \
   --prompt_data OpenRLHF/prompt-collection-v0.1 \
   --input_key context_messages \
   --apply_chat_template \
   --max_samples 25000 \
   --normalize_reward \
   --adam_offload \
   --flash_attn \
   --load_checkpoint \
   --gradient_checkpointing \
   --n_samples_per_prompt 8 \
   --freezing_actor_steps 10 \
   --wandb_run_name ppo_64-8x8_1203_01
```
